otherlibs/systhreads/st_stubs.c

```diff
@@ -377,8 +377,9 @@ static void caml_thread_reinitialize(void)
      the effective owner of the lock. So there is no need to run
      st_masterlock_acquire (busy = 1) */
   st_masterlock *m = Thread_lock(Caml_state->id);
-  m->init = 0; /* force initialization */
-  st_masterlock_init(m);
+  m->init = 0; /* force reinitialization */
+  if (0 != st_masterlock_init(m))
+    caml_fatal_error("Unix.fork: failed to reinitialize master lock");
```

Comment:
I can see that the proposed change is better than what was earlier. Is there no better recovery here than a `caml_fatal_error`?

Reply:
Is even caml_fatal_error safe here, given the earlier giant "only execute async-signal-safe operations" comment in the diff hunk above?   If we've failed to do something as simple as allocate a few mutexes at this point in the fork, just exiting is probably the only option.

Reply:
Indeed if it fails, there is not much that can be done at that point. In case there is something to improve to the "best effort" to support fork+threads, it is still better to attempt writing an error message. Note that when trying to support fork+threads, one is already deep in unchartered territory (with some hope that the global lock ensures enough consistency to let you ignore the "only async-signal-safe" requirement). Of course, the whole enterprise is doomed to failure if there are other C threads and one is interrupted in the middle of malloc, say.